doc/manual/faq.rst

```diff
@@ -196,6 +196,24 @@ create and use variable-length arrays in kernels?
+
+See also :ref:`basic-artiq-python`. Let's run down the options for kernel-host data transfer:
+
+    - Kernels can return single values directly. They *cannot* return lists, arrays or strings, because of the way these values are allocated, which prevents values of these types from outliving the kernel they are created in. This is still true when the values in question are wrapped in functions or objects, in which case they may be missed by lifetime tracking and accepted by the compiler, but will cause memory corruption when run.
+
+    - Kernels can freely make changes to attributes of objects shared with the host, including ``self``. However, these changes will be made to a kernel-owned copy of the object, which is only synchronized with the host copy when the kernel completes. This means that host-side operations executed during the runtime of the kernel, including RPCs, will be handling an unmodified version of the object, and modifications made by those operations will simply be overwritten when the kernel returns.
+
+    .. note::
+        Attribute writeback happens *once per kernel*, that is, if your experiment contains many separate kernels called from the host, modifications will be written back when each separate kernel completes. This is generally not suitable for data transfer, however, as new kernels are costly to create, and experiments often try to avoid doing so. It is also important to specify that kernels called *from* a kernel will not write back to the host upon completion. Attribute writeback is only executed upon return to the host.
+
+    - Kernels can interact with datasets, either as attributes (if :meth:`~artiq.language.environment.HasEnvironment.setattr_dataset` is used) or by RPC of the get and set methods (:meth:`~artiq.language.environment.HasEnvironment.get_dataset`, :meth:`~artiq.language.environment.HasEnvironment.set_dataset`, etc.). In this case note that, like certain other host-side methods, :meth:`~artiq.language.environment.HasEnvironment.get_dataset` will not actually be accepted by the compiler, because its return type is not specified. To call it as an RPC, simply wrap it in another function which *does* specify a return type. :meth:`~artiq.language.environment.HasEnvironment.set_dataset` can be similarly wrapped to make it asynchronous.
+
+    - Kernels can of course also call arbitrary RPCs. When sending data to the host, these can be asynchronous, and this is normally the recommended way of transferring data back to the host, resulting in a relatively minor amount of delay in the kernel. When receiving data from the host, they must be synchronous, which is still considerably faster than the overhead of compiling and starting a new kernel. Note that if data is being both (asynchronously) sent and received, the RPCs may execute in parallel on the host side, and minor race conditions can result (i.e. retrieved data may not yet show updates sent in an earlier RPC).
```

Comment:
Regular and async RPCs are still guaranteed to arrive at the host in-order. In other words, when encountering a regular RPC call, all outstanding async RPC requests will be sent first.

Reply:
Side note: have we documented the conditions under which async RPCs can become blocking?

Reply:
I should have known better than to try to answer from memory: now that I have it all in front of me, it is in fact mentioned [here](https://m-labs.hk/artiq/manual/compiler.html#asynchronous-rpcs), just not in very great detail. But regardless the current phrasing here almost implies async RPCs never block and I'll fix that in a moment. 

> Regular and async RPCs are still guaranteed to arrive at the host in-order. In other words, when encountering a regular RPC call, all outstanding async RPC requests will be sent first.

*Sent* yes, but not necessarily *executed?* Unless that's been specifically compensated for since 2020? I'm going by sb10q's response in [this forum post](https://forum.m-labs.hk/d/72-get-data-asynchronously/6).

Reply:
I don't understand what @sbourdeauducq is getting at in that forum post. Both the sending and receiving side should be fully serial (the kernel CPU is, and the worker is synchronous as well).

We are relying on the consistent ordering all over the place (e.g. in ndscan) to get low-overhead synchronisation between kernel and host. If there is a bug breaking this, that would be concerning (but also, I haven't seen any behaviour like that in practice).